prometheus/inc/class-plugin.php

```diff
@@ -157,14 +92,71 @@ private static function create_registry(): RegistryInterface {
 		return new CollectorRegistry( $storage );
 	}
 
+	public function get_collectors(): array {
+		return $this->collectors;
+	}
+
 	/**
-	 * Validate if current request is for the Prometheus endpoint.
+	 * Cron callback to process the metrics for collection
 	 *
-	 * @return bool
+	 * @return void
+	 */
+	public function process_metrics(): void {
+		if ( Context::is_web_request() ) {
+			trigger_error( __METHOD__ . ' should not be called on web requests', E_USER_WARNING );
+			return;
+		}
+
+		foreach ( $this->collectors as $collector ) {
+			if ( is_callable( [ $collector, 'process_metrics' ] ) ) {
+				$collector->process_metrics();
+			}
+		}
+	}
+
+	/**
+	 * We're going to send the response to the client, then collect metrics from each registered collector
 	 */
-	private function is_prom_endpoint_request(): bool {
-		// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized -- the value is used only for strict comparison
-		$request_uri = $_SERVER['REQUEST_URI'] ?? '';
-		return self::$endpoint_path === $request_uri;
+	public function shutdown(): void {
+		// This is expensive, potentially, so be mindful about when this method is called
+		// Currently it only runs on web requests with a 90 interval, see the constructor
+		if ( function_exists( 'fastcgi_finish_request' ) ) {
+			fastcgi_finish_request();
+		}
+
+		// Don't run on unauthorized requests.
+		if ( ! function_exists( 'wp_get_current_user' ) || ! current_user_can( 'edit_posts' ) ) {
+			return;
+		}
```

Comment:
Not sure I understand the `! function_exists( 'wp_get_current_user' )`. And then for the latter, it's quite possible a request from a user with `edit_posts` permissions doesn't occur for a pretty long period of time, is that intended/okay to go so long without collecting?

Reply:
I don't remember in which context exactly but I got `wp_get_current_user` fatal so I added this check. 

`collect_metrics` is something that grabs the result of computationally-intensive of `process_metrics` (which happens off request path`. It doesn't really matter when it's exactly called, see the Posts Stats collector for the example implementation.

Reply:
The intent of this whole method is purely to limit the situations where metrics can be collected and when the single job event is scheduled.